File: platform/js/splitter.js
/* =============================================================================
   AI4Mobility — Exercise Step Splitter
   -----------------------------------------------------------------------------
   Takes a free-text exercise description and splits it into ordered, actionable
   steps.

   HONEST STATEMENT OF METHOD (this matters — see docs/report):
   This is a RULE-BASED linguistic splitter. It is NOT a large language model.
   No model weights are loaded and no network call is made. Everything below is
   deterministic, inspectable and runs offline in the browser.

   An LLM adapter interface is defined at the bottom of this file
   (`LLMSplitterAdapter`). It is a STUB: it is not connected to any provider and
   returns null. It exists so the swap is a one-file change, and so the report
   can describe exactly what would be required. Do not present it as working.

   Pipeline
   --------
     1. normalise()        clean whitespace, unify quotes/dashes, expand "&"
     2. detectExplicit()   if the author already numbered or bulleted the steps,
                           trust that and stop — never re-split a numbered list
     3. segment()          sentence segmentation with abbreviation protection
     4. splitCompound()    break sentences on coordinating connectives, but only
                           when BOTH sides look like instructions
     5. classify()         label each fragment: action | position | hold | reps |
                           safety | context
     6. extractParams()    pull hold seconds and repetition counts out of text
     7. score()            per-step confidence + an overall confidence
     8. assemble()         drop non-instructional fragments, order, renumber
   ============================================================================= */

(function (global) {
  'use strict';

  /* ---------------------------------------------------------------- lexicon */

  // Verbs that begin an instruction. Ordered roughly by frequency in the NHS
  // and physiotherapy leaflets used as source material.
  const ACTION_VERBS = [
    'lie', 'lay', 'sit', 'stand', 'kneel', 'start', 'begin', 'come', 'get',
    'place', 'put', 'position', 'rest', 'set',
    'bend', 'straighten', 'extend', 'flex', 'stretch', 'reach', 'raise',
    'lift', 'lower', 'drop', 'push', 'pull', 'press', 'squeeze', 'tighten',
    'relax', 'release', 'hold', 'keep', 'maintain',
    'move', 'slide', 'bring', 'take', 'turn', 'rotate', 'twist', 'roll',
    'point', 'tuck', 'curl', 'round', 'arch', 'fold', 'open', 'close',
    'breathe', 'inhale', 'exhale',
    'repeat', 'return', 'continue', 'swap', 'switch', 'change',
    'walk', 'step', 'rise', 'stay', 'pause', 'stop', 'avoid', 'ensure', 'make'
  ];

  // Phrases that mark a resting/starting position rather than a movement.
  const POSITION_MARKERS = [
    'starting position', 'start position', 'position:', 'starting:',
    'begin in', 'start in', 'from a', 'in the position'
  ];

  // Phrases that mark safety guidance rather than a step to perform.
  const SAFETY_MARKERS = [
    'stop if', 'seek advice', 'do not', "don't", 'avoid', 'caution',
    'if pain', 'if it hurts', 'should not', 'never ', 'consult',
    'healthcare professional', 'physiotherapist', 'contraindic'
  ];

  // Coordinators we are willing to split a sentence on.
  const SPLIT_CONNECTIVES = [
    ', then ', ' then ', ', and then ', ' and then ',
    ', next ', ' next, ', ', after that ', ' after that ',
    ', followed by ', '; '
  ];

  // Abbreviations that must not end a sentence.
  const ABBREVIATIONS = [
    'e.g.', 'i.e.', 'etc.', 'approx.', 'dr.', 'mr.', 'mrs.', 'ms.',
    'st.', 'no.', 'vs.', 'fig.', 'sec.', 'min.', 'reps.'
  ];

  /* -------------------------------------------------------------- normalise */

  function normalise(text) {
    if (typeof text !== 'string') return '';
    return text
      .replace(/\r\n?/g, '\n')
      .replace(/[‘’]/g, "'")
      .replace(/[“”]/g, '"')
      .replace(/[–—]/g, '-')
      .replace(/ /g, ' ')
      .replace(/\s*&\s*/g, ' and ')
      .replace(/[ \t]+/g, ' ')
      .replace(/\n{3,}/g, '\n\n')
      .trim();
  }

  /* --------------------------------------------------- explicit list detect */

  // If the author has already numbered or bulleted their steps, splitting again
  // does more harm than good. Detect and preserve that structure.
  function detectExplicit(text) {
    const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
    if (lines.length < 2) return null;

    const NUMBERED = /^(\d{1,2})\s*[.)\]:-]\s+(.{3,})$/;
    const BULLETED = /^[-*•‣◦⁃]\s+(.{3,})$/;

    const numbered = [];
    const bulleted = [];
    lines.forEach(l => {
      const n = l.match(NUMBERED);
      if (n) numbered.push(n[2].trim());
      const b = l.match(BULLETED);
      if (b) bulleted.push(b[1].trim());
    });

    // Require the marked lines to be the clear majority of the input.
    if (numbered.length >= 2 && numbered.length >= lines.length * 0.6) {
      return { items: numbered, kind: 'numbered' };
    }
    if (bulleted.length >= 2 && bulleted.length >= lines.length * 0.6) {
      return { items: bulleted, kind: 'bulleted' };
    }
    return null;
  }

  /* ---------------------------------------------------------- segmentation */

  function segment(text) {
    const protectedText = text.replace(/\b(\d+)\.(\d+)/g, '$1<DEC>$2');
    let working = protectedText;
    ABBREVIATIONS.forEach((abbr, i) => {
      const re = new RegExp(abbr.replace(/\./g, '\\.'), 'gi');
      working = working.replace(re, `<ABBR${i}>`);
    });

    const parts = working
      .split(/(?<=[.!?])\s+(?=[A-Z0-9])|\n+/)
      .map(s => s.trim())
      .filter(Boolean);

    return parts.map(s => {
      let out = s.replace(/<DEC>/g, '.');
      ABBREVIATIONS.forEach((abbr, i) => {
        out = out.replace(new RegExp(`<ABBR${i}>`, 'g'), abbr);
      });
      return out.trim();
    }).filter(s => s.length > 2);
  }

  /* ------------------------------------------------------- compound split */

  function startsWithAction(fragment) {
    const first = fragment.trim().toLowerCase().split(/[^a-z']+/).filter(Boolean)[0];
    if (!first) return false;
    if (ACTION_VERBS.includes(first)) return true;
    // "Slowly lift ..." — allow one leading adverb.
    const words = fragment.trim().toLowerCase().split(/[^a-z']+/).filter(Boolean);
    if (words.length > 1 && /ly$/.test(words[0]) && ACTION_VERBS.includes(words[1])) return true;
    return false;
  }

  function splitCompound(sentence) {
    // Try each connective; only accept the split if the right-hand side reads
    // as its own instruction. This is what stops "lift your leg and hold it
    // level" from becoming two useless fragments.
    for (const conn of SPLIT_CONNECTIVES) {
      const idx = sentence.toLowerCase().indexOf(conn);
      if (idx > 8) {
        const left = sentence.slice(0, idx).trim().replace(/[,;]$/, '');
        const right = sentence.slice(idx + conn.length).trim();
        if (left.length > 8 && right.length > 8 && startsWithAction(right)) {
          return [left, ...splitCompound(right)];
        }
      }
    }
    return [sentence];
  }

  /* ------------------------------------------------------------- classify */

  function classify(fragment) {
    const low = fragment.toLowerCase();

    if (SAFETY_MARKERS.some(m => low.includes(m))) return 'safety';
    if (POSITION_MARKERS.some(m => low.includes(m))) return 'position';
    if (/^(hold|keep|maintain|stay)\b/.test(low)) return 'hold';
    if (/^(repeat|do this|perform)\b/.test(low) || /\brepeat\s+\d+\s+times\b/.test(low)) return 'reps';
    if (startsWithAction(fragment)) return 'action';
    if (/^(lie|lay|sit|stand|kneel|start|begin)\b/.test(low)) return 'position';
    return 'context';
  }

  /* --------------------------------------------------------- parameters */

  const NUM_WORDS = {
    one: 1, two: 2, three: 3, four: 4, five: 5, six: 6, seven: 7,
    eight: 8, nine: 9, ten: 10, eleven: 11, twelve: 12, fifteen: 15,
    twenty: 20, thirty: 30, sixty: 60
  };

  function toNumber(token) {
    if (token == null) return null;
    const t = String(token).toLowerCase().trim();
    if (/^\d+$/.test(t)) return parseInt(t, 10);
    if (NUM_WORDS[t] != null) return NUM_WORDS[t];
    return null;
  }

  function extractParams(fragment) {
    const low = fragment.toLowerCase();
    const params = { hold_s: null, reps: null, minutes: null };

    // hold for 5 seconds / hold for five seconds / hold 5-10 seconds
    let m = low.match(/hold(?:ing)?(?:\s+(?:it|this|the position))?\s+(?:for\s+)?(\d+|[a-z]+)(?:\s*(?:-|to)\s*(\d+|[a-z]+))?\s*(?:seconds?|secs?|s)\b/);
    if (m) {
      const a = toNumber(m[1]);
      const b = toNumber(m[2]);
      if (a != null) params.hold_s = b != null ? Math.round((a + b) / 2) : a;
    }

    // lie for 30 minutes
    m = low.match(/(?:for\s+)?(\d+|[a-z]+)\s*(?:minutes?|mins?)\b/);
    if (m) {
      const a = toNumber(m[1]);
      if (a != null) params.minutes = a;
    }

    // repeat 10 times / do this ten times / 10 repetitions / x10
    m = low.match(/(\d+|[a-z]+)\s*(?:times|reps|repetitions)\b/);
    if (m) {
      const a = toNumber(m[1]);
      if (a != null && a <= 100) params.reps = a;
    }
    if (params.reps == null) {
      m = low.match(/\bx\s*(\d{1,3})\b/);
      if (m) params.reps = parseInt(m[1], 10);
    }

    return params;
  }

  /* ------------------------------------------------------------- scoring */

  function scoreStep(fragment, kind) {
    let s = 0.4;
    if (kind === 'action') s += 0.35;
    else if (kind === 'position') s += 0.25;
    else if (kind === 'hold' || kind === 'reps') s += 0.2;

    const words = fragment.split(/\s+/).length;
    if (words >= 4 && words <= 28) s += 0.15;      // sensible instruction length
    else if (words > 40) s -= 0.2;                  // probably still compound

    if (/\b(your|the)\b/.test(fragment.toLowerCase())) s += 0.05;  // addresses a body part
    if (/[,;]\s*(and|then)\b/i.test(fragment)) s -= 0.1;           // unsplit compound

    return Math.max(0.05, Math.min(0.99, s));
  }

  /* ------------------------------------------------------------- assemble */

  function splitExerciseText(rawText, options) {
    const opts = Object.assign({ keepSafety: true, minConfidence: 0.35 }, options || {});
    const text = normalise(rawText);

    if (!text) {
      return { ok: false, reason: 'empty', steps: [], safety: [], confidence: 0, method: 'none' };
    }
    if (text.split(/\s+/).length < 4) {
      return { ok: false, reason: 'too_short', steps: [], safety: [], confidence: 0, method: 'none' };
    }

    let fragments;
    let method;

    const explicit = detectExplicit(text);
    if (explicit) {
      fragments = explicit.items;
      method = 'explicit-' + explicit.kind;
    } else {
      fragments = [];
      segment(text).forEach(sentence => {
        splitCompound(sentence).forEach(f => fragments.push(f));
      });
      method = 'rule-based';
    }

    const safety = [];
    const steps = [];

    fragments.forEach(frag => {
      const clean = frag.replace(/^[,;:\-\s]+/, '').replace(/[,;:\s]+$/, '').trim();
      if (clean.length < 4) return;

      const kind = classify(clean);
      if (kind === 'safety') {
        if (opts.keepSafety) safety.push(sentenceCase(clean));
        return;
      }

      const conf = scoreStep(clean, kind);
      if (kind === 'context' && conf < opts.minConfidence) return;

      const params = extractParams(clean);
      steps.push({
        n: 0,
        text: sentenceCase(clean),
        kind: kind,
        hold_s: params.hold_s || (params.minutes ? params.minutes * 60 : 0),
        reps: params.reps || null,
        confidence: Math.round(conf * 100) / 100
      });
    });

    steps.forEach((s, i) => { s.n = i + 1; });

    const confidence = steps.length
      ? Math.round((steps.reduce((a, s) => a + s.confidence, 0) / steps.length) * 100) / 100
      : 0;

    return {
      ok: steps.length > 0,
      reason: steps.length ? 'ok' : 'no_steps_found',
      method: method,
      steps: steps,
      safety: safety,
      confidence: confidence,
      warnings: buildWarnings(steps, confidence)
    };
  }

  function buildWarnings(steps, confidence) {
    const w = [];
    if (steps.length === 1) {
      w.push('Only one step was found. If the description has more, try putting each instruction on its own line.');
    }
    if (steps.length > 15) {
      w.push('That produced a lot of steps. Check whether some should be merged.');
    }
    if (confidence < 0.55 && steps.length) {
      w.push('Low confidence — please read the steps and correct anything that looks wrong.');
    }
    const longOnes = steps.filter(s => s.text.split(/\s+/).length > 35).length;
    if (longOnes) {
      w.push(longOnes + ' step(s) are quite long and may still contain more than one instruction.');
    }
    return w;
  }

  function sentenceCase(s) {
    const t = s.trim();
    if (!t) return t;
    let out = t.charAt(0).toUpperCase() + t.slice(1);
    if (!/[.!?]$/.test(out)) out += '.';
    return out;
  }

  /* --------------------------------------------------------- LLM adapter */

  /**
   * STUB — NOT IMPLEMENTED.
   *
   * This is the seam where a language model would replace the rule-based
   * splitter. It is deliberately left unimplemented rather than faked.
   *
   * To implement it you would need to:
   *   1. Choose a provider and hold an API key server-side (never in this file —
   *      the platform is a static page and any key here would be public).
   *   2. Add a small backend endpoint that proxies the request, so the key stays
   *      on the server and requests can be rate-limited.
   *   3. Send the description with a schema-constrained prompt asking for an
   *      ordered array of {text, hold_s, reps}.
   *   4. Validate the returned JSON against that schema before display, and keep
   *      the rule-based splitter as the fallback when validation fails.
   *   5. Review the clinical-safety implications: a model may invent an
   *      instruction that was not in the source description. Any LLM output
   *      shown to a parent would need the original text displayed beside it,
   *      which is why the interface already does that.
   *
   * Until those five things are done, this returns null and the caller falls
   * back to splitExerciseText().
   */
  const LLMSplitterAdapter = {
    available: false,
    reason: 'Not implemented. No model, no API key, no backend endpoint. See comments in splitter.js.',
    async split(/* rawText */) {
      return null;
    }
  };

  /** Public entry point. Tries the LLM adapter, falls back to rules. */
  async function splitExercise(rawText, options) {
    if (LLMSplitterAdapter.available) {
      try {
        const viaLLM = await LLMSplitterAdapter.split(rawText);
        if (viaLLM && Array.isArray(viaLLM.steps) && viaLLM.steps.length) {
          return Object.assign({ method: 'llm' }, viaLLM);
        }
      } catch (e) {
        /* fall through to rules */
      }
    }
    return splitExerciseText(rawText, options);
  }

  const api = {
    splitExercise,
    splitExerciseText,
    LLMSplitterAdapter,
    _internal: { normalise, segment, splitCompound, classify, extractParams, detectExplicit, scoreStep }
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  global.AI4MSplitter = api;

})(typeof window !== 'undefined' ? window : globalThis);

File: src/splitter.py
"""
AI4Mobility — Exercise Step Splitter (Python mirror).

A faithful port of ``platform/js/splitter.js`` so the notebook can run the same
splitting logic that the web interface uses, and so both can be tested against
the same expectations.

HONEST STATEMENT OF METHOD
--------------------------
This is a RULE-BASED linguistic splitter. It is not a large language model. No
model weights are loaded and no network call is made. Everything here is
deterministic and inspectable.

An LLM adapter seam is provided at the bottom (``LLMSplitterAdapter``). It is a
stub that returns ``None``; it is not connected to any provider. See the module
docstring there for exactly what implementing it would require.

Usage
-----
    from src.splitter import split_exercise

    result = split_exercise(
        "Lay on your back with your leg straight. Point your foot and toes "
        "down, then slowly pull your foot up towards your knee. Repeat 10 times."
    )
    for step in result.steps:
        print(step.n, step.text, step.reps, step.hold_s)
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field, asdict
from typing import Any, Optional

__all__ = [
    "Step", "SplitResult", "split_exercise", "split_exercise_text",
    "LLMSplitterAdapter",
]

# --------------------------------------------------------------------- lexicon

ACTION_VERBS = {
    "lie", "lay", "sit", "stand", "kneel", "start", "begin", "come", "get",
    "place", "put", "position", "rest", "set",
    "bend", "straighten", "extend", "flex", "stretch", "reach", "raise",
    "lift", "lower", "drop", "push", "pull", "press", "squeeze", "tighten",
    "relax", "release", "hold", "keep", "maintain",
    "move", "slide", "bring", "take", "turn", "rotate", "twist", "roll",
    "point", "tuck", "curl", "round", "arch", "fold", "open", "close",
    "breathe", "inhale", "exhale",
    "repeat", "return", "continue", "swap", "switch", "change",
    "walk", "step", "rise", "stay", "pause", "stop", "avoid", "ensure", "make",
}

POSITION_MARKERS = (
    "starting position", "start position", "position:", "starting:",
    "begin in", "start in", "from a", "in the position",
)

SAFETY_MARKERS = (
    "stop if", "seek advice", "do not", "don't", "avoid", "caution",
    "if pain", "if it hurts", "should not", "never ", "consult",
    "healthcare professional", "physiotherapist", "contraindic",
)

SPLIT_CONNECTIVES = (
    ", then ", " then ", ", and then ", " and then ",
    ", next ", " next, ", ", after that ", " after that ",
    ", followed by ", "; ",
)

ABBREVIATIONS = (
    "e.g.", "i.e.", "etc.", "approx.", "dr.", "mr.", "mrs.", "ms.",
    "st.", "no.", "vs.", "fig.", "sec.", "min.", "reps.",
)

NUM_WORDS = {
    "one": 1, "two": 2, "three": 3, "four": 4, "five": 5, "six": 6,
    "seven": 7, "eight": 8, "nine": 9, "ten": 10, "eleven": 11, "twelve": 12,
    "fifteen": 15, "twenty": 20, "thirty": 30, "sixty": 60,
}


# --------------------------------------------------------------- data classes

@dataclass
class Step:
    n: int
    text: str
    kind: str
    hold_s: int = 0
    reps: Optional[int] = None
    confidence: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class SplitResult:
    ok: bool
    reason: str
    method: str
    steps: list[Step] = field(default_factory=list)
    safety: list[str] = field(default_factory=list)
    confidence: float = 0.0
    warnings: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        d = asdict(self)
        d["steps"] = [s.to_dict() for s in self.steps]
        return d

    def __len__(self) -> int:
        return len(self.steps)


# ------------------------------------------------------------------ normalise

def normalise(text: Optional[str]) -> str:
    if not isinstance(text, str):
        return ""
    out = text.replace("\r\n", "\n").replace("\r", "\n")
    out = out.replace("‘", "'").replace("’", "'")
    out = out.replace("“", '"').replace("”", '"')
    out = re.sub(r"[–—]", "-", out)
    out = out.replace(" ", " ")
    out = re.sub(r"\s*&\s*", " and ", out)
    out = re.sub(r"[ \t]+", " ", out)
    out = re.sub(r"\n{3,}", "\n\n", out)
    return out.strip()


# ------------------------------------------------------- explicit list detect

_NUMBERED = re.compile(r"^(\d{1,2})\s*[.)\]:-]\s+(.{3,})$")
_BULLETED = re.compile(r"^[-*•‣◦⁃]\s+(.{3,})$")


def detect_explicit(text: str) -> Optional[tuple[list[str], str]]:
    """Preserve a list the author already numbered or bulleted."""
    lines = [ln.strip() for ln in text.split("\n") if ln.strip()]
    if len(lines) < 2:
        return None

    numbered, bulleted = [], []
    for ln in lines:
        m = _NUMBERED.match(ln)
        if m:
            numbered.append(m.group(2).strip())
        m = _BULLETED.match(ln)
        if m:
            bulleted.append(m.group(1).strip())

    if len(numbered) >= 2 and len(numbered) >= len(lines) * 0.6:
        return numbered, "numbered"
    if len(bulleted) >= 2 and len(bulleted) >= len(lines) * 0.6:
        return bulleted, "bulleted"
    return None


# --------------------------------------------------------------- segmentation

def segment(text: str) -> list[str]:
    working = re.sub(r"\b(\d+)\.(\d+)", r"\1<DEC>\2", text)
    for i, abbr in enumerate(ABBREVIATIONS):
        working = re.sub(re.escape(abbr), f"<ABBR{i}>", working, flags=re.IGNORECASE)

    parts = re.split(r"(?<=[.!?])\s+(?=[A-Z0-9])|\n+", working)

    out: list[str] = []
    for part in parts:
        s = part.strip()
        if not s:
            continue
        s = s.replace("<DEC>", ".")
        for i, abbr in enumerate(ABBREVIATIONS):
            s = s.replace(f"<ABBR{i}>", abbr)
        s = s.strip()
        if len(s) > 2:
            out.append(s)
    return out


# ------------------------------------------------------------- compound split

def _words(fragment: str) -> list[str]:
    return [w for w in re.split(r"[^a-z']+", fragment.strip().lower()) if w]


def starts_with_action(fragment: str) -> bool:
    ws = _words(fragment)
    if not ws:
        return False
    if ws[0] in ACTION_VERBS:
        return True
    # allow one leading adverb: "Slowly lift ..."
    return len(ws) > 1 and ws[0].endswith("ly") and ws[1] in ACTION_VERBS


def split_compound(sentence: str) -> list[str]:
    low = sentence.lower()
    for conn in SPLIT_CONNECTIVES:
        idx = low.find(conn)
        if idx > 8:
            left = sentence[:idx].strip().rstrip(",;").strip()
            right = sentence[idx + len(conn):].strip()
            if len(left) > 8 and len(right) > 8 and starts_with_action(right):
                return [left] + split_compound(right)
    return [sentence]


# ------------------------------------------------------------------- classify

def classify(fragment: str) -> str:
    low = fragment.lower()
    if any(m in low for m in SAFETY_MARKERS):
        return "safety"
    if any(m in low for m in POSITION_MARKERS):
        return "position"
    if re.match(r"^(hold|keep|maintain|stay)\b", low):
        return "hold"
    if re.match(r"^(repeat|do this|perform)\b", low) or re.search(r"\brepeat\s+\d+\s+times\b", low):
        return "reps"
    if starts_with_action(fragment):
        return "action"
    if re.match(r"^(lie|lay|sit|stand|kneel|start|begin)\b", low):
        return "position"
    return "context"


# ----------------------------------------------------------------- parameters

def _to_number(token: Optional[str]) -> Optional[int]:
    if token is None:
        return None
    t = str(token).strip().lower()
    if t.isdigit():
        return int(t)
    return NUM_WORDS.get(t)


_RE_HOLD = re.compile(
    r"hold(?:ing)?(?:\s+(?:it|this|the position))?\s+(?:for\s+)?"
    r"(\d+|[a-z]+)(?:\s*(?:-|to)\s*(\d+|[a-z]+))?\s*(?:seconds?|secs?|s)\b"
)
_RE_MIN = re.compile(r"(?:for\s+)?(\d+|[a-z]+)\s*(?:minutes?|mins?)\b")
_RE_REPS = re.compile(r"(\d+|[a-z]+)\s*(?:times|reps|repetitions)\b")
_RE_X = re.compile(r"\bx\s*(\d{1,3})\b")


def extract_params(fragment: str) -> dict[str, Optional[int]]:
    low = fragment.lower()
    params: dict[str, Optional[int]] = {"hold_s": None, "reps": None, "minutes": None}

    m = _RE_HOLD.search(low)
    if m:
        a, b = _to_number(m.group(1)), _to_number(m.group(2))
        if a is not None:
            params["hold_s"] = round((a + b) / 2) if b is not None else a

    m = _RE_MIN.search(low)
    if m:
        a = _to_number(m.group(1))
        if a is not None:
            params["minutes"] = a

    m = _RE_REPS.search(low)
    if m:
        a = _to_number(m.group(1))
        if a is not None and a <= 100:
            params["reps"] = a
    if params["reps"] is None:
        m = _RE_X.search(low)
        if m:
            params["reps"] = int(m.group(1))

    return params


# -------------------------------------------------------------------- scoring

def score_step(fragment: str, kind: str) -> float:
    s = 0.4
    if kind == "action":
        s += 0.35
    elif kind == "position":
        s += 0.25
    elif kind in ("hold", "reps"):
        s += 0.20

    n_words = len(fragment.split())
    if 4 <= n_words <= 28:
        s += 0.15
    elif n_words > 40:
        s -= 0.20

    if re.search(r"\b(your|the)\b", fragment.lower()):
        s += 0.05
    if re.search(r"[,;]\s*(and|then)\b", fragment, flags=re.IGNORECASE):
        s -= 0.10

    return max(0.05, min(0.99, s))


# ------------------------------------------------------------------- assemble

def _sentence_case(s: str) -> str:
    t = s.strip()
    if not t:
        return t
    out = t[0].upper() + t[1:]
    if not re.search(r"[.!?]$", out):
        out += "."
    return out


def _build_warnings(steps: list[Step], confidence: float) -> list[str]:
    w: list[str] = []
    if len(steps) == 1:
        w.append("Only one step was found. If the description has more, try "
                 "putting each instruction on its own line.")
    if len(steps) > 15:
        w.append("That produced a lot of steps. Check whether some should be merged.")
    if confidence < 0.55 and steps:
        w.append("Low confidence — please read the steps and correct anything "
                 "that looks wrong.")
    long_ones = sum(1 for s in steps if len(s.text.split()) > 35)
    if long_ones:
        w.append(f"{long_ones} step(s) are quite long and may still contain "
                 f"more than one instruction.")
    return w


def split_exercise_text(raw_text: Optional[str], keep_safety: bool = True,
                        min_confidence: float = 0.35) -> SplitResult:
    """Split a free-text exercise description into ordered steps."""
    text = normalise(raw_text)

    if not text:
        return SplitResult(False, "empty", "none")
    if len(text.split()) < 4:
        return SplitResult(False, "too_short", "none")

    explicit = detect_explicit(text)
    if explicit:
        fragments, kind = explicit
        method = f"explicit-{kind}"
    else:
        fragments = []
        for sentence in segment(text):
            fragments.extend(split_compound(sentence))
        method = "rule-based"

    safety: list[str] = []
    steps: list[Step] = []

    for frag in fragments:
        clean = frag.strip(" ,;:-\t").strip()
        if len(clean) < 4:
            continue

        kind = classify(clean)
        if kind == "safety":
            if keep_safety:
                safety.append(_sentence_case(clean))
            continue

        conf = score_step(clean, kind)
        if kind == "context" and conf < min_confidence:
            continue

        params = extract_params(clean)
        hold = params["hold_s"] or ((params["minutes"] or 0) * 60)
        steps.append(Step(
            n=0,
            text=_sentence_case(clean),
            kind=kind,
            hold_s=int(hold),
            reps=params["reps"],
            confidence=round(conf, 2),
        ))

    for i, s in enumerate(steps):
        s.n = i + 1

    confidence = round(sum(s.confidence for s in steps) / len(steps), 2) if steps else 0.0

    return SplitResult(
        ok=bool(steps),
        reason="ok" if steps else "no_steps_found",
        method=method,
        steps=steps,
        safety=safety,
        confidence=confidence,
        warnings=_build_warnings(steps, confidence),
    )


# --------------------------------------------------------------- LLM adapter

class LLMSplitterAdapter:
    """
    STUB — NOT IMPLEMENTED.

    The seam where a language model would replace the rule-based splitter. It is
    deliberately left unimplemented rather than faked.

    Implementing it requires:
      1. A provider and an API key held server-side.
      2. A backend endpoint proxying the request so the key never reaches the
         client, with rate limiting.
      3. A schema-constrained prompt returning an ordered array of
         ``{text, hold_s, reps}``.
      4. Validation of the returned JSON before display, with this rule-based
         splitter kept as the fallback on validation failure.
      5. A clinical safety review: a model may invent an instruction that was
         not present in the source description. This is why the interface shows
         the original text beside the generated steps.

    Until those are done, ``split`` returns ``None``.
    """

    available: bool = False
    reason: str = ("Not implemented. No model, no API key, no backend endpoint. "
                   "See the class docstring.")

    @staticmethod
    def split(raw_text: str) -> Optional[SplitResult]:  # noqa: ARG004
        return None


def split_exercise(raw_text: Optional[str], **kwargs: Any) -> SplitResult:
    """Public entry point: try the LLM adapter, fall back to the rules."""
    if LLMSplitterAdapter.available:
        try:
            via_llm = LLMSplitterAdapter.split(raw_text or "")
            if via_llm and via_llm.steps:
                via_llm.method = "llm"
                return via_llm
        except Exception:  # noqa: BLE001 — never let the optional path break the fallback
            pass
    return split_exercise_text(raw_text, **kwargs)


if __name__ == "__main__":  # pragma: no cover
    demo = ("Lay on your back with your leg straight. Point your foot and toes "
            "down to the floor or bed, then slowly pull your foot and toes up "
            "towards your knee. Repeat 10 times. Stop if pain becomes worse.")
    res = split_exercise(demo)
    print(f"method={res.method}  confidence={res.confidence}")
    for s in res.steps:
        extra = []
        if s.reps:
            extra.append(f"{s.reps}x")
        if s.hold_s:
            extra.append(f"{s.hold_s}s hold")
        print(f"  {s.n}. [{s.kind}] {s.text}" + (f"   ({', '.join(extra)})" if extra else ""))
    for w in res.safety:
        print(f"  ! safety: {w}")
